src/IB/IBFESurfaceMethod.cpp

```diff
@@ -1406,111 +2110,944 @@ IBFESurfaceMethod::initializeLevelData(Pointer<BasePatchHierarchy<NDIM> > hierar
+                                         << " the pressure interpolation ghost width hasn't beeen properly set"
+                                         << std::endl);
+        }
+
+        const unsigned int nindices = static_cast<int>(local_indices.size());
+
+
+        if (!local_indices.empty())
+        {
+            for (unsigned int k = 0; k < nindices; ++k)
+            {
+                if (d_calculate_interior_side)
+                    P_qp[local_indices[k]] = -(P_jump_qp[local_indices[k]] - Q_io_qp[local_indices[k]]);
+                else
+                    P_qp[local_indices[k]] =
+                        P_jump_qp[local_indices[k]] +
+                        Q_io_qp[local_indices[k]]; // 0.5 * ( P_jump_qp[local_indices[k]] + Q_i_qp[local_indices[k]] +
+                                                   // Q_o_qp[local_indices[k]]);
+            }
+        }
+
+        // Loop over the elements and accumulate the right-hand-side values.
+        qrule.reset();
+        qp_offset = 0;
+        for (unsigned int e_idx = 0; e_idx < num_active_patch_elems; ++e_idx)
+        {
+            Elem* const elem = patch_elems[e_idx];
+
+            P_dof_map_cache.dof_indices(elem, P_dof_indices);
+            P_rhs_e.resize(static_cast<int>(P_dof_indices.size()));
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                X_dof_map.dof_indices(elem, X_dof_indices[d], d);
+            }
+            get_values_for_interpolation(x_node, *X_ghost_vec, X_dof_indices);
+            const bool qrule_changed =
+                FEDataManager::updateInterpQuadratureRule(qrule, d_default_interp_spec, elem, x_node, patch_dx_min);
+            if (qrule_changed) fe->attach_quadrature_rule(qrule.get());
+
+            fe->reinit(elem);
+
+            const unsigned int n_qp = qrule->n_points();
+            const size_t n_basis = X_dof_indices[0].size();
+            for (unsigned int qp = 0; qp < n_qp; ++qp)
+            {
+                const int idx = qp_offset + qp;
+                for (unsigned int k = 0; k < n_basis; ++k)
+                {
+                    const double p_JxW = phi[k][qp] * JxW[qp];
+                    P_rhs_e(k) += P_qp[idx] * p_JxW;
+                }
+            }
+
+            P_dof_map.constrain_element_vector(P_rhs_e, P_dof_indices);
+            P_rhs_vec->add_vector(P_rhs_e, P_dof_indices);
+
+            qp_offset += n_qp;
+        }
+    }
+    P_rhs_vec->close();
+
+    d_fe_data_managers[part]->computeL2Projection(
+        *P_vec, *P_rhs_vec, P_SYSTEM_NAME, d_default_interp_spec.use_consistent_mass_matrix);
+
+    d_X_half_vecs[part]->close();
+    d_X_current_vecs[part]->close();
+    d_X_new_vecs[part]->close();
+
+    d_P_half_vecs[part]->close();
+    d_P_jump_IB_ghost_vecs[part]->close();
+    d_X_IB_ghost_vecs[part]->close();
+
+    return;
+
+} // interpolatePressureForTraction
+
+void
+IBFESurfaceMethod::computeFluidTraction(const double data_time, unsigned int part)
```

Comment:
Same comment as above: Suggest making the flag about interior/exterior side an argument to the function instead of a class member variable.